File: sdks/java/aether-sdk/src/test/java/io/aether/sdk/streaming/PartitionTest.java
package io.aether.sdk.streaming;

import io.aether.sdk.streaming.Types.StreamEvent;
import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;

import java.util.List;

class PartitionTest {

    @Test
    @DisplayName("hash partition returns valid index")
    void testHashPartition() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.HASH;
        config.partitions = 4;
        Partitioner partitioner = new Partitioner(config);
        int p = partitioner.partitionByKey("user-123");
        assertTrue(p >= 0 && p < 4);
    }

    @Test
    @DisplayName("round robin distributes across partitions")
    void testRoundRobin() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.ROUND_ROBIN;
        config.partitions = 3;
        Partitioner partitioner = new Partitioner(config);
        int p1 = partitioner.partitionByKey("key");
        int p2 = partitioner.partitionByKey("key");
        int p3 = partitioner.partitionByKey("key");
        assertNotEquals(p1, p2);
        assertNotEquals(p2, p3);
    }

    @Test
    @DisplayName("null key falls back to round robin")
    void testNullKeyFallback() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.HASH;
        config.partitions = 5;
        Partitioner partitioner = new Partitioner(config);
        int p = partitioner.partitionByKey(null);
        assertTrue(p >= 0 && p < 5);
    }

    @Test
    @DisplayName("empty key falls back to round robin")
    void testEmptyKeyFallback() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.HASH;
        config.partitions = 5;
        Partitioner partitioner = new Partitioner(config);
        int p = partitioner.partitionByKey("");
        assertTrue(p >= 0 && p < 5);
    }

    @Test
    @DisplayName("random partition returns valid index")
    void testRandomPartition() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.RANDOM;
        config.partitions = 10;
        Partitioner partitioner = new Partitioner(config);
        for (int i = 0; i < 100; i++) {
            int p = partitioner.partitionByKey("key-" + i);
            assertTrue(p >= 0 && p < 10);
        }
    }

    @Test
    @DisplayName("partition by event")
    void testPartitionByEvent() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.KEY;
        config.partitions = 4;
        Partitioner partitioner = new Partitioner(config);
        StreamEvent<String> event = StreamEvent.create("my-key", "value");
        int p = partitioner.partition(event);
        assertTrue(p >= 0 && p < 4);
    }

    @Test
    @DisplayName("partitionByValue with key extractor")
    void testPartitionByValue() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.KEY;
        config.partitions = 3;
        config.keyExtractor = v -> ((String) v).substring(0, 1);
        Partitioner partitioner = new Partitioner(config);
        int p = partitioner.partitionByValue("abc");
        assertTrue(p >= 0 && p < 3);
    }

    @Test
    @DisplayName("getNumPartitions returns correct count")
    void testGetNumPartitions() {
        PartitionConfig config = new PartitionConfig();
        config.partitions = 7;
        Partitioner partitioner = new Partitioner(config);
        assertEquals(7, partitioner.getNumPartitions());
    }

    @Test
    @DisplayName("zero partitions defaults to 1")
    void testZeroPartitionsDefaults() {
        PartitionConfig config = new PartitionConfig();
        config.partitions = 0;
        Partitioner partitioner = new Partitioner(config);
        assertEquals(1, partitioner.getNumPartitions());
    }

    @Test
    @DisplayName("getStats tracks distribution")
    void testGetStats() {
        PartitionConfig config = new PartitionConfig();
        config.strategy = PartitionStrategy.KEY;
        config.partitions = 3;
        Partitioner partitioner = new Partitioner(config);
        partitioner.partitionByKey("a");
        partitioner.partitionByKey("b");
        PartitionStats stats = partitioner.getStats();
        assertEquals(2, stats.totalEvents.get());
    }

    @Test
    @DisplayName("rebalance resets stats")
    void testRebalance() {
        PartitionConfig config = new PartitionConfig();
        config.partitions = 3;
        Partitioner partitioner = new Partitioner(config);
        partitioner.partitionByKey("a");
        partitioner.rebalance(5);
        assertEquals(5, partitioner.getNumPartitions());
        PartitionStats stats = partitioner.getStats();
        assertEquals(0, stats.totalEvents.get());
        assertEquals(1, stats.rebalances.get());
    }

    @Test
    @DisplayName("KeyExtractor extracts key")
    void testKeyExtractor() {
        KeyExtractor<String> extractor = new KeyExtractor<>(s -> s.split(":")[0]);
        assertEquals("user", extractor.extract("user:123"));
    }

    @Test
    @DisplayName("KeyExtractor fallback for null result")
    void testKeyExtractorFallback() {
        KeyExtractor<String> extractor = new KeyExtractor<>(s -> null);
        assertEquals("default", extractor.extract("anything"));
        extractor.setFallback("fallback");
        assertEquals("fallback", extractor.extract("anything"));
    }

    @Test
    @DisplayName("KeyExtractor stats")
    void testKeyExtractorStats() {
        KeyExtractor<String> extractor = new KeyExtractor<>(s -> s);
        extractor.extract("a");
        extractor.extract(null);
        long[] stats = extractor.getStats();
        assertEquals(2, stats[0]);
        assertEquals(1, stats[1]);
    }

    @Test
    @DisplayName("PartitionProcessor tracks events and errors")
    void testPartitionProcessor() throws Exception {
        PartitionProcessor<String> pp = new PartitionProcessor<>(0, event -> null);
        pp.process(StreamEvent.create("k", "v"));
        long[] stats = pp.getStats();
        assertEquals(1, stats[0]);
        assertEquals(0, stats[1]);
        assertEquals(0, pp.getPartitionId());
    }

    @Test
    @DisplayName("PartitionProcessor error counting")
    void testPartitionProcessorError() {
        PartitionProcessor<String> pp = new PartitionProcessor<>(0, event -> {
            throw new RuntimeException("fail");
        });
        assertThrows(Exception.class, () -> pp.process(StreamEvent.create("k", "v")));
        assertEquals(1, pp.getStats()[1]);
    }

    @Test
    @DisplayName("PartitionStrategy enum values")
    void testPartitionStrategyValues() {
        assertEquals(5, PartitionStrategy.values().length);
        assertNotNull(PartitionStrategy.valueOf("ROUND_ROBIN"));
        assertNotNull(PartitionStrategy.valueOf("KEY"));
        assertNotNull(PartitionStrategy.valueOf("HASH"));
        assertNotNull(PartitionStrategy.valueOf("RANDOM"));
        assertNotNull(PartitionStrategy.valueOf("RANGE"));
    }
}

File: sdks/go/aether/streaming/partition_test.go
package streaming

import (
	"testing"
)

func TestPartitionStrategy_String(t *testing.T) {
	tests := []struct {
		ps   PartitionStrategy
		want string
	}{
		{PartitionStrategyRoundRobin, "RoundRobin"},
		{PartitionStrategyKey, "Key"},
		{PartitionStrategyHash, "Hash"},
		{PartitionStrategyRandom, "Random"},
		{PartitionStrategyRange, "Range"},
		{PartitionStrategy(99), "Unknown"},
	}
	for _, tt := range tests {
		if got := tt.ps.String(); got != tt.want {
			t.Errorf("PartitionStrategy(%d).String() = %q, want %q", tt.ps, got, tt.want)
		}
	}
}

func TestPartitioner_New(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyKey,
		NumPartitions: 10,
	})
	if p.NumPartitions() != 10 {
		t.Errorf("expected 10, got %d", p.NumPartitions())
	}
}

func TestPartitioner_New_DefaultPartitions(t *testing.T) {
	p := NewPartitioner(PartitionConfig{})
	if p.NumPartitions() != 1 {
		t.Errorf("expected default 1 partition, got %d", p.NumPartitions())
	}
}

func TestPartitioner_PartitionByKey(t *testing.T) {
	tests := []struct {
		name     string
		strategy PartitionStrategy
		key      string
		want     int
	}{
		{"key strategy", PartitionStrategyKey, "user-1", 0},
		{"hash strategy", PartitionStrategyHash, "user-1", 0},
		{"round robin", PartitionStrategyRoundRobin, "any", 0},
		{"empty key round robin", PartitionStrategyKey, "", 0},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := NewPartitioner(PartitionConfig{
				Strategy:      tt.strategy,
				NumPartitions: 10,
			})
			part := p.PartitionByKey(tt.key)
			if part < 0 || part >= 10 {
				t.Errorf("partition %d out of range [0, 10)", part)
			}
		})
	}
}

func TestPartitioner_PartitionByKey_Consistent(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyKey,
		NumPartitions: 10,
	})

	p1 := p.PartitionByKey("consistent-key")
	p2 := p.PartitionByKey("consistent-key")
	if p1 != p2 {
		t.Error("same key should always route to same partition")
	}
}

func TestPartitioner_PartitionByKey_Different(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyKey,
		NumPartitions: 10,
	})

	p1 := p.PartitionByKey("key-a")
	p2 := p.PartitionByKey("key-b")
	if p1 == p2 {
		t.Error("different keys should (likely) route to different partitions")
	}
}

func TestPartitioner_Partition(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyKey,
		NumPartitions: 5,
	})

	event := NewStreamEvent[any]("user-1", "data")
	part := p.Partition(event)
	if part < 0 || part >= 5 {
		t.Errorf("partition %d out of range [0, 5)", part)
	}
}

func TestPartitioner_PartitionByValue(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy: PartitionStrategyKey,
		KeyExtractor: func(v any) string {
			return "extracted"
		},
		NumPartitions: 5,
	})

	part := p.PartitionByValue("anything")
	if part < 0 || part >= 5 {
		t.Errorf("partition %d out of range", part)
	}
}

func TestPartitioner_PartitionByValue_NoExtractor(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyRoundRobin,
		NumPartitions: 3,
	})

	p1 := p.PartitionByValue("a")
	p2 := p.PartitionByValue("b")
	if p1 == p2 {
		t.Error("round robin should give different partitions")
	}
}

func TestPartitioner_RangePartition(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyRange,
		NumPartitions: 3,
		RangeBounds:   []string{"m", "s"},
	})

	p1 := p.PartitionByKey("a")
	p2 := p.PartitionByKey("n")
	p3 := p.PartitionByKey("z")

	if p1 >= p2 {
		t.Error("'a' should be in earlier partition than 'n'")
	}
	if p2 >= p3 {
		t.Error("'n' should be in earlier partition than 'z'")
	}
}

func TestPartitioner_RangePartition_NoBounds(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyRange,
		NumPartitions: 5,
		RangeBounds:   nil,
	})

	part := p.PartitionByKey("any")
	if part != 0 {
		t.Errorf("expected 0 for no bounds, got %d", part)
	}
}

func TestPartitioner_GetStats(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyKey,
		NumPartitions: 5,
	})

	p.PartitionByKey("k1")
	p.PartitionByKey("k2")
	p.PartitionByKey("k3")

	stats := p.GetStats()
	if stats.TotalEvents != 3 {
		t.Errorf("expected 3 total events, got %d", stats.TotalEvents)
	}
	if len(stats.PartitionCount) != 5 {
		t.Errorf("expected 5 partition counts, got %d", len(stats.PartitionCount))
	}
}

func TestPartitioner_Rebalance(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyKey,
		NumPartitions: 3,
	})

	p.PartitionByKey("k1")
	p.Rebalance(10)

	if p.NumPartitions() != 10 {
		t.Errorf("expected 10 after rebalance, got %d", p.NumPartitions())
	}
	stats := p.GetStats()
	if stats.Rebalances != 1 {
		t.Errorf("expected 1 rebalance, got %d", stats.Rebalances)
	}
}

func TestKeyExtractor_New(t *testing.T) {
	ke := NewKeyExtractor[string](func(s string) string {
		return s
	})

	if ke.Extract("test-key") != "test-key" {
		t.Errorf("expected 'test-key', got %q", ke.Extract("test-key"))
	}
}

func TestKeyExtractor_Empty(t *testing.T) {
	ke := NewKeyExtractor[string](func(s string) string {
		return ""
	})

	if ke.Extract("test") != "default" {
		t.Errorf("expected 'default' fallback, got %q", ke.Extract("test"))
	}
}

func TestKeyExtractor_Nil(t *testing.T) {
	ke := NewKeyExtractor[string](nil)
	if ke.Extract("test") != "default" {
		t.Errorf("expected 'default' for nil extractor, got %q", ke.Extract("test"))
	}
}

func TestKeyExtractor_SetFallback(t *testing.T) {
	ke := NewKeyExtractor[string](nil)
	ke.SetFallback("fallback")
	if ke.Extract("test") != "fallback" {
		t.Errorf("expected 'fallback', got %q", ke.Extract("test"))
	}
}

func TestKeyExtractor_Stats(t *testing.T) {
	ke := NewKeyExtractor[string](func(s string) string { return s })
	ke.Extract("a")
	ke.Extract("b")
	ke.Extract("")

	count, nullCount := ke.Stats()
	if count != 3 {
		t.Errorf("expected 3 extractions, got %d", count)
	}
	if nullCount != 1 {
		t.Errorf("expected 1 null, got %d", nullCount)
	}
}

func TestPartitionProcessor(t *testing.T) {
	var eventCount, errorCount int64
	pp := NewPartitionProcessor[int](1, func(event StreamEvent[int]) error {
		eventCount++
		return nil
	})

	pp.Process(NewStreamEvent[int]("k", 42))
	pp.Process(NewStreamEvent[int]("k", 43))

	if pp.PartitionID() != 1 {
		t.Errorf("expected partition ID 1, got %d", pp.PartitionID())
	}

	ec, errCnt := pp.Stats()
	if ec != 2 {
		t.Errorf("expected 2 events, got %d", ec)
	}
	if errCnt != 0 {
		t.Errorf("expected 0 errors, got %d", errCnt)
	}
	_ = eventCount
	_ = errorCount
}

func TestPartitionProcessor_Error(t *testing.T) {
	pp := NewPartitionProcessor[int](0, func(event StreamEvent[int]) error {
		return &BatchError{Message: "fail"}
	})

	pp.Process(NewStreamEvent[int]("k", 1))
	_, errCnt := pp.Stats()
	if errCnt != 1 {
		t.Errorf("expected 1 error, got %d", errCnt)
	}
}

func TestCompositePartitioner_New(t *testing.T) {
	cp := NewCompositePartitioner[string](
		[]PartitionStrategy{PartitionStrategyKey, PartitionStrategyHash},
		[]float64{0.7, 0.3},
		[]func(string) string{func(s string) string { return s }, nil},
		5,
	)

	if cp.NumPartitions() != 5 {
		t.Errorf("expected 5, got %d", cp.NumPartitions())
	}

	part := cp.Partition("test-key")
	if part < 0 || part >= 5 {
		t.Errorf("partition %d out of range", part)
	}
}

func TestCompositePartitioner_AddStrategy(t *testing.T) {
	cp := NewCompositePartitioner[string](
		[]PartitionStrategy{PartitionStrategyKey},
		[]float64{1.0},
		[]func(string) string{func(s string) string { return s }},
		5,
	)

	cp.AddStrategy(PartitionStrategyHash, 0.5, func(s string) string { return s })
	part := cp.Partition("key")
	if part < 0 || part >= 5 {
		t.Errorf("partition %d out of range", part)
	}
}

func TestDefaultPartitionConfig(t *testing.T) {
	cfg := DefaultPartitionConfig()
	if cfg.Strategy != PartitionStrategyKey {
		t.Errorf("expected key strategy, got %v", cfg.Strategy)
	}
	if cfg.NumPartitions != 10 {
		t.Errorf("expected 10, got %d", cfg.NumPartitions)
	}
}

func TestPartitioner_RoundRobinDistribution(t *testing.T) {
	p := NewPartitioner(PartitionConfig{
		Strategy:      PartitionStrategyRoundRobin,
		NumPartitions: 3,
	})

	seen := make(map[int]bool)
	for i := 0; i < 30; i++ {
		part := p.PartitionByKey("any")
		seen[part] = true
	}

	if len(seen) < 2 {
		t.Error("round robin should distribute across partitions")
	}
}
